### PR TITLE
make project title editable, both in standalone and www modes

### PR DESCRIPTION
**Pull request by benjiwheeler**
Big help from pairing with @LiFaytheGoblin!!

relies on https://github.com/LLK/scratch-www/pull/2052 and should only be merged after it.

### Resolves

https://github.com/LLK/scratch-gui/issues/2815

Note: we haven't implemented the css changes per @carljbowman's comments on #2815.

### Proposed Changes

Makes project title editable in GUI editor. 

Uses new project title reducer to store and change project title. This is made available to the menu bar, where the project title input uses it to maintain state, and to project saver, which uses the title to determine the downloaded filename.

When run within www, title changes made in the GUI menu bar are updated in www and through it to the server.

Testing:

For testing, we should try unusual unicode characters.

### Related PR

https://github.com/LLK/scratch-www/pull/2052 in scratch-www should be used together with this.

### CODE REVIEW

File: src/components/menu-bar/menu-bar.jsx [hunk 1]
```diff
@@ -152,6 +155,26 @@ class MenuBar extends React.Component {
             this.props.onRequestCloseEdit();
         };
     }
+    // intercept key presses to detect enter key; make that trigger blur,
+    // and thus project title update
+    handleProjectTitleKeyPress (e) {
```
[the remark this answers is not in the document]
Comment (rschamp) [reply src/components/menu-bar/menu-bar.jsx]: [BufferedInputHOC](https://github.com/LLK/scratch-gui/blob/develop/src/components/forms/buffered-input-hoc.jsx) for reference.

**Comment by LiFaytheGoblin** [comment]
CSS is at https://github.com/LiFaytheGoblin/scratch-gui/tree/rename_title

Tested for MacOS: Chrome, Safari, Firefox

**Comment by LiFaytheGoblin** [comment]
Cool looks like this branch here got updated as well, so nevermind my repo link!

**Comment by benjiwheeler** [comment]
@rschamp, we were thinking that we *do* want to always enable editing the title field -- if embedded in www, obviously; but also in beta, as the title for the file when it downloads.

File: src/components/menu-bar/menu-bar.jsx [hunk 2]
```diff
@@ -355,11 +378,17 @@ class MenuBar extends React.Component {
...
+                        >
                             <input
-                                disabled
                                 className={classNames(styles.titleField)}
-                                placeholder="Untitled-1"
+                                placeholder=""
+                                value={this.props.projectTitle}
+                                onBlur={this.handleUpdateProjectTitle}
+                                onChange={this.handleProjectTitleChange}
+                                onKeyPress={this.handleProjectTitleKeyPress}
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/menu-bar/menu-bar.jsx]: Done!

File: src/components/menu-bar/menu-bar.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/menu-bar/menu-bar.jsx]: Done!

File: src/reducers/project-title.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+const SET_PROJECT_TITLE = 'projectTitle/SET_PROJECT_TITLE';
+
+const projectTitleInitialState = {
+    projectTitle: 'Untitled-1'
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/reducers/project-title.js]: Done!

File: src/containers/gui.jsx [hunk 4]
```diff
@@ -100,12 +109,19 @@ GUI.propTypes = {
...
     previewInfoVisible: PropTypes.bool,
     projectData: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
     projectHost: PropTypes.string,
+    projectTitle: PropTypes.string,
     vm: PropTypes.instanceOf(VM)
 };
 
+GUI.defaultProps = {
+    projectTitle: 'Untitled-1'
+};
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/containers/gui.jsx]: Done!

**Comment by benjiwheeler** [comment]
Linda and I confirmed that although several tests fail, when run individually, they pass.

**Comment by LiFaytheGoblin** [comment]
Note for changes to css: Don't use width = outline but box shadow:
border: none;
    border-radius: calc(calc(0.5rem / 2) * 2);
    box-shadow: 0 0 0 0.25rem hsla(0, 100%, 100%, 0.25);

**Comment by LiFaytheGoblin** [comment]
one more css thing

File: src/components/menu-bar/project-title-input.css [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+.title-field:hover {
+    background-color: hsla(0, 100%, 100%, 0.5);
+}
+
+.title-field:focus {
+    outline:none;
+    border: none;
+    border-radius: calc($form-radius * 2); /* double the border width */
```
Comment (LiFaytheGoblin) [inline src/components/menu-bar/project-title-input.css]: border radius line needs to be removed

File: src/containers/gui.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (fsih) [reply src/containers/gui.jsx]: It looks like this is still here?

File: src/containers/project-saver.jsx [hunk 6]
```diff
@@ -33,11 +33,7 @@ class ProjectSaver extends React.Component {
         document.body.appendChild(saveLink);
 
         this.props.saveProjectSb3().then(content => {
-            // TODO user-friendly project name
-            // File name: project-DATE-TIME
-            const date = new Date();
-            const timestamp = `${date.toLocaleDateString()}-${date.toLocaleTimeString()}`;
-            const filename = `untitled-project-${timestamp}.sb3`;
+            const filename = `${this.props.projectTitle}.sb3`;
```
Comment (fsih) [inline src/containers/project-saver.jsx]: The behavior is kind of weird if the title field is blank (downloads sb3.zip)

File: src/lib/project-loader-hoc.jsx [hunk 7]
```diff
@@ -72,7 +72,6 @@ const ProjectLoaderHOC = function (WrappedComponent) {
                 assetHost,
                 projectHost,
                 projectId,
-                reduxProjectId,
```
Comment (fsih) [inline src/lib/project-loader-hoc.jsx]: Why is this removed?

File: src/containers/project-saver.jsx (hunk 6 above)
Comment (benjiwheeler) [reply src/containers/project-saver.jsx]: Hm, good point. I'll give it a default in that case.

File: src/lib/project-loader-hoc.jsx (hunk 7 above)
Comment (benjiwheeler) [reply src/lib/project-loader-hoc.jsx]: The line is a bug. It was left there by Ray by accident. Sorry, I should have pointed that out!

File: src/components/menu-bar/project-title-input.css (hunk 5 above)
Comment (benjiwheeler) [reply src/components/menu-bar/project-title-input.css]: Done

**Comment by benjiwheeler** [comment]
@fsih I made a default title if it's blank, and set a maxLength and tested it on Safari. Thanks for catching those bugs.

File: src/components/menu-bar/project-title-input.jsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    constructor (props) {
+        super(props);
+        bindAll(this, [
+            'handleUpdateProjectTitle'
+        ]);
+    }
+    // call onUpdateProjectTitle if it is defined (only defined when gui
+    // is used within scratch-www)
+    handleUpdateProjectTitle (newTitle) {
+        this.props.onSetProjectTitle(newTitle);
```
Comment (rschamp) [inline src/components/menu-bar/project-title-input.jsx]: I think it'd be more correct if this were something like:
```jsx
        if (this.props.onUpdateProjectTitle) {
            this.props.onUpdateProjectTitle(newTitle);
        } else {
            this.props.onSetProjectTitle(newTitle);
        }
```

That way the redux state won't get out of sync from the incoming prop, and changing the title would work without configuration.

When `onUpdateProjectTitle` is supplied, calling it should cause the parent component (www or the gui playground) to pass in a new title prop to the GUI component, and then that change should dispatch the redux state change, causing everything looking at the redux state to update.

Comment (rschamp) [reply src/components/menu-bar/project-title-input.jsx]: I also think there's an issue with the naming here. It's not clear what the difference is between `onUpdateProjectTitle` and `onUpdateProjectTitle`. I think it's because they're the same thing. If we move to the behavior I outlined above, then I think it would be possible to collapse these into a single prop.

File: src/playground/index.jsx [hunk 9]
```diff
@@ -28,6 +28,13 @@ document.body.appendChild(appTarget);
 GUI.setAppElement(appTarget);
 const WrappedGui = HashParserHOC(AppStateHOC(GUI));
 
+// simple example of how you might manage project title externally.
+// Changing project title within GUI interface will update it here.
+let projectTitle = 'Untitled-1';
+const onUpdateProjectTitle = title => {
+    projectTitle = title;
```
Comment (rschamp) [inline src/playground/index.jsx]: Does this really cause the GUI to re-render? I would expect this to happen silently, so you would need to have a wrapper component that held this in state (like in the player example).

**Comment by rschamp** [comment]
Also, projects on scratch.mit.edu must have titles 100 characters or shorter. I guess that doesn't matter in standalone mode, but seems like a sensible cap if we want to add one into the title field here.

File: src/components/menu-bar/project-title-input.jsx (hunk 8 above)
Comment (benjiwheeler) [reply src/components/menu-bar/project-title-input.jsx]: Done.

File: src/playground/index.jsx (hunk 9 above)
Comment (benjiwheeler) [reply src/playground/index.jsx]: NO it doesn't! Fixed.

File: src/playground/index.jsx [hunk 10]
```diff
@@ -26,7 +27,39 @@ appTarget.className = styles.app;
...
+// simple example of how you might manage project title externally.
+// Changing project title within GUI interface will update it here.
+class TitledGUI extends React.Component {
+    constructor (props) {
+        super(props);
+        bindAll(this, [
+            'handleUpdateProjectTitle'
+        ]);
+        this.state = {
+            projectTitle: 'Untitled-1'
```
Comment (rschamp) [inline src/playground/index.jsx]: Would you please update this default to "Scratch Project"?  This came up in a planning meeting, and is only relevant here, in the standalone playground, not scratch-www.

Also, this string should use `formatMessage` so it can be translated.

**Comment by benjiwheeler** [comment]
@rschamp, please take a look at the titled-hoc.jsx and where it's used in player.jsx, playground/index.jsx, and render-gui.jsx

File: src/playground/index.jsx [hunk 11]
```diff
@@ -27,7 +28,7 @@ if (supportedBrowser()) {
 
 } else {
     BrowserModalComponent.setAppElement(appTarget);
-    const WrappedBrowserModalComponent = AppStateHOC(BrowserModalComponent, true /* localesOnly */);
+    const WrappedBrowserModalComponent = AppStateHOC(TitledHOC(BrowserModalComponent, true /* localesOnly */));
```
Comment (rschamp) [inline src/playground/index.jsx]: You shouldn't wrap this one, since it's just the unsupported browser modal. There's also an issue with parenthesis placement (`true` should be passed to AppStateHOC, not TitledHOC), which shows that we should change the API for AppStateHOC to be used like `AppStateHOC(true)(Component))` rather than the kind of odd and hard to read `AppStateHOC(Component, true)`.

File: src/containers/project-loader.jsx [hunk 12]
```diff
@@ -75,6 +76,12 @@ class ProjectLoader extends React.Component {
         if (thisFileInput.files) { // Don't attempt to load if no file was selected
             this.props.openLoadingState();
             reader.readAsArrayBuffer(thisFileInput.files[0]);
+            if (thisFileInput.files[0].name) {
+                const matches = thisFileInput.files[0].name.match(/^(.*)\.sb3$/);
```
Comment (rschamp) [inline src/containers/project-loader.jsx]: You could match the first 100 characters here to be consistent with the cap elsewhere

File: src/containers/project-loader.jsx [hunk 13]
```diff
@@ -75,6 +76,12 @@ class ProjectLoader extends React.Component {
...
+                if (matches) {
+                    this.props.onSetProjectTitle(matches[1]);
```
Comment (rschamp) [inline src/containers/project-loader.jsx]: Or just truncate to 100 here.

File: src/playground/index.jsx (hunk 11 above)
Comment (benjiwheeler) [reply src/playground/index.jsx]: Sorry, the parenthesis was sloppy of me! Should have caught that.

Comment (benjiwheeler) [reply src/playground/index.jsx]: Done.

File: src/containers/project-loader.jsx (hunk 12 above)
Comment (benjiwheeler) [reply src/containers/project-loader.jsx]: Done

File: src/containers/project-loader.jsx (hunk 13 above)
Comment (benjiwheeler) [reply src/containers/project-loader.jsx]: Done
